### PR TITLE
[LI-HOTFIX] Prevent preferred controller set to over-shrink in controlled shutdown process

### PR DESCRIPTION
**Pull request by lmr3796**
LI_DESCRIPTION = LIKAFKA-38857
EXIT_CRITERIA = N/A

Similar to the ISR check in controlled shutdown, this patch offers the check on the preferred controllers during controlled shutdown.

A new config `li.min.preferred.controller.count` is introduced to specify the threshold.
If shutting down a broker would cause the preferred controller count to drop below the threshold, the controller would reject the controlled shutdown requests with corresponding the error code `NOT_ENOUGH_PREFERRED_CONTROLLERS`; this applies to shutting down the active controller itself.

Also added an integration test case.

### CODE REVIEW

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 1]
```diff
@@ -1733,16 +1733,31 @@ class KafkaController(val config: KafkaConfig,
       else brokerEpoch
 
     val shouldSkipShutdownSafetyCheck = controllerContext.skipShutdownSafetyCheck.getOrElse(id, -1L) >= actualBrokerEpoch
+    var passedAllShutdown: Boolean = true // For logging purpose
+    if (controllerContext.getLivePreferredControllerIds.size <= config.liMinPreferredControllerCount
+        // Only check if this broker affects live preferred controller set so that for non-preferred-controller clusters it doesn't get blocked
```
Comment (CCisGG) [inline core/src/main/scala/kafka/controller/KafkaController.scala:1738]: Didn't quite get how this check can rule out non-prefere-controller clusters. Could you elaborate a bit?

Comment (CCisGG) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1738]: Do you mean for non-prefered-controller clusters, the liMinPreferredControllerCount is set to a very small value like 0?

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 2]
```diff
@@ -1733,16 +1733,31 @@ class KafkaController(val config: KafkaConfig,
...
+        && controllerContext.getLivePreferredControllerIds.contains(id)) {
+      passedAllShutdown &= false
```
Comment (CCisGG) [inline core/src/main/scala/kafka/controller/KafkaController.scala]: Why do you need this? And why not directly set it to false? Looks like  passedAllShutdown &= false will always set the value to false.

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 3]
```diff
@@ -82,6 +82,7 @@ object Defaults {
   val MetadataSnapshotMaxNewRecordBytes = 20 * 1024 * 1024
   val ProducerBatchDecompressionEnable = true
   val PreferredController = false
+  val LiMinPreferredControllerCount = 0
```
Comment (CCisGG) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:85]: I see. So by default the cluster is treated as non-preferred-controller cluster. We will need to explicitly set this config value to enable the protection.

**Review by CCisGG — COMMENTED** [review]
Left some questions.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 2 above)
Comment (lmr3796) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: That's a good catch!

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 3 above)
Comment (lmr3796) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:85]: Yes that's the intent.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 1 above)
Comment (lmr3796) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1738]: >  liMinPreferredControllerCount is set to a very small value like 0?

That's a more of an orthogonal angle.



Only preferred controller would be in `controllerContext.getLivePreferredControllerIds`.  So if I'm shutting down either
1. data broker
2. a controller in a cluster that doesn't configure preferred controller
3. a "fallback" controller (i.e. preferred controllers are down, but this cluster allows data broker to take controller for availability reason)

Those would not fall in the to the if block.

Comment (CCisGG) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1738]: Sounds good. Thanks for the clarification!

**Review by CCisGG — APPROVED** [review]

**Pull request merged**